lib/Doctrine/DBAL/Driver/Mysqli/MysqliStatement.php

```diff
@@ -151,59 +192,87 @@ public function execute($params = null)
-            // if mysqli is compiled with libmysql, subsequently fetched string values will get truncated
-            // to the length of the ones fetched during the previous execution.
-            $this->_rowBindedValues = array_fill(0, count($this->_columnNames), null);
-
-            $refs = [];
-            foreach ($this->_rowBindedValues as $key => &$value) {
-                $refs[$key] =& $value;
-            }
+        // Store result of every execution which has it. Otherwise it will be impossible
+        // to execute a new statement in case if the previous one has non-fetched rows
+        // @link http://dev.mysql.com/doc/refman/5.7/en/commands-out-of-sync.html
+        $this->_stmt->store_result();
+
+        // Bind row values _after_ storing the result. Otherwise, if mysqli is compiled with libmysql,
+        // it will have to allocate as much memory as it may be needed for the given column type
+        // (e.g. for a LONGBLOB field it's 4 gigabytes)
+        // @link https://bugs.php.net/bug.php?id=51386#1270673122
+        //
+        // Make sure that the values are bound after each execution. Otherwise, if closeCursor() has been
+        // previously called on the statement, the values are unbound making the statement unusable.
+        //
+        // It's also important that row values are bound after _each_ call to store_result(). Otherwise,
+        // if mysqli is compiled with libmysql, subsequently fetched string values will get truncated
+        // to the length of the ones fetched during the previous execution.
+        assert(is_array($this->_columnNames));
+        $this->_rowBindedValues = array_fill(0, count($this->_columnNames), null);
 
-            if (! $this->_stmt->bind_result(...$refs)) {
-                throw new MysqliException($this->_stmt->error, $this->_stmt->sqlstate, $this->_stmt->errno);
-            }
+        $refs = [];
+        foreach ($this->_rowBindedValues as $key => &$value) {
+            $refs[$key] =& $value;
+        }
+
+        if (! $this->_stmt->bind_result(...$refs)) {
+            throw new MysqliException($this->_stmt->error, $this->_stmt->sqlstate, $this->_stmt->errno);
         }
 
         $this->result = true;
 
         return true;
     }
 
+    /**
+     * Converts an array of named parameters, e.g. ['id' => 1, 'foo' => 'bar'] to the corresponding array with
+     * positional parameters referring to the prepared query, e.g. [1 => 1, 2 => 'bar', 3 => 'bar'] for a prepared query
+     * like "SELECT id FROM table WHERE foo = :foo and baz = :foo".
+     *
+     * @param array<int|string, mixed>|null $params
+     *
+     * @return mixed[]|null more specific: array<int, mixed>, I just don't know an elegant way to convince phpstan
+     */
+    private function convertNamedToPositionalParamsIfNeeded(?array $params = null)
+    {
+        if ($params === null || count($params) === 0) {
```

Comment:
The case of `[]` does not need to be singled out IMO.
```suggestion
        if ($params === null) {
```

Reply:
The idea was to deal with any kind of emptyness first, but the empty array behaves well in the following code, too. So I'm happy to go either way.